Profile popup interface

wesleypayton opened the pull request:
## Describe your changes
- Added singleton popup service
- Added popup menu to social page
- Popup menu buttons navigate to pages and close popup after navigation
- Removed pets page button
- Changed routing so that back navigation works properly

## Issue ticket number or link
Fixes https://github.com/ChrispyPeaches/FocusFriends/issues/25

## Documentation
Documentation will be added to the frontend design section below
https://github.com/ChrispyPeaches/FocusFriends/wiki/Frontend-Design

### Please describe the tests that you ran to verify your changes.
 - [x] Verified applications built successfully
 - [x] Popup menu opens and closes properly
 - [x] Popup buttons work properly
 - [x] Navigation animations still work properly
 - [x] Pages and buttons do not cause errors

wesleypayton commented:
I changed how the routing works so that back navigation works properly now. Hitting the back button on the settings page will direct the user back to the timer or social page depending on which page the user came from. As a side effect of this change, the tab bar disappears when navigating into the menu.

Please play around in the branch locally and let me know how yall feel about this change as well as the look of the popup menu.

ChrispyPeaches commented on src/FocusApp.Client/MauiProgram.cs:
```diff
@@ -14,6 +14,8 @@
 using SimpleToolkit.SimpleShell;
 using Microsoft.EntityFrameworkCore.Migrations.Internal;
 using FocusApp.Client.Views.Shop;
+using FocusApp.Client.Helpers;
```
Could we remove this repeated reference?

ChrispyPeaches commented on src/FocusApp.Client/MauiProgram.cs:
```diff
@@ -40,6 +42,8 @@ public static MauiApp CreateMauiApp()
                 .RegisterServices()
                 .RegisterPages();
 
+            builder.Services.AddTransient<ProfilePopupInterface>();
```
Since we're expected to have multiple popups throughout the app, and each will be registered as a transient, do you think it would be helpful to put this into a method? It could mimic the RegisteredServices method where it registers each individually, or it could mimic RegisterPages and have a BasePopup class. 

What are your thoughts on these suggestions?

wesleypayton replied:
I added a BasePopup class as suggested as well as a RegisterPopups function similar to the RegisterPages. Good shout 🙌

ChrispyPeaches commented on src/FocusApp.Client/MauiProgram.cs:
```diff
@@ -71,6 +75,7 @@ private static IServiceCollection RegisterServices(this IServiceCollection servi
         {
             // Registered as a singleton so the timer is not reset by page navigation
             services.AddSingleton<ITimerService, TimerService>();
+            services.AddSingleton<Helpers.PopupService>();
```
Since there's an interface already made for the PopupService, could we have its injection mimic the <ITimerService, TimerService> injection for consistency?

ChrispyPeaches commented on src/FocusApp.Client/Helpers/PopupService.cs:
```diff
@@ -0,0 +1,50 @@
+
+    public class PopupService : IPopupService
+    {
+        private readonly IServiceProvider services;
+
+        // Need to keep track of the popups so we can close them.
+        private Stack<Popup> popups = new();
```
Could we have these fields renamed to match the c# standards? (just have an _ at the beginning of the name, there are some examples in TimerService, but it would be _popups and _services.

Additionally, could we have both of them instantiated in the constructor? 
It helps with readability and maintainability

wesleypayton replied:
I renamed the variables but what do you mean by instantiated in the constructor?

ChrispyPeaches reviewed (changes requested):
Overall looks really good and is extremely visually appealing!
I can tell you put a lot of time into the visuals and I really appreciate that!
I also really appreciate how clean the `ProfilePopupInterface.PageButtonClicked()` method is!

I did request some changes, however they're all very minor. 
Really great work here!

ChrispyPeaches commented:
> I changed how the routing works so that back navigation works properly now. Hitting the back button on the settings page will direct the user back to the timer or social page depending on which page the user came from. As a side effect of this change, the tab bar disappears when navigating into the settings/pets menu.
> 
> Please play around in the branch locally and let me know how yall feel about this change as well as the look of the popup menu.

About this, I think it makes sense that the settings page doesn't have the tab bar, but I think it would make more sense to have the tab bar on the Pets page. Since the user will likely want to see how each look on their island and be able to quickly switch back and forth.

Would this change be possible?

wesleypayton commented:
> @wesleypayton
> 
> > I changed how the routing works so that back navigation works properly now. Hitting the back button on the settings page will direct the user back to the timer or social page depending on which page the user came from. As a side effect of this change, the tab bar disappears when navigating into the settings/pets menu.
> > Please play around in the branch locally and let me know how yall feel about this change as well as the look of the popup menu.
> 
> About this, I think it makes sense that the settings page doesn't have the tab bar, but I think it would make more sense to have the tab bar on the Pets page. Since the user will likely want to see how each look on their island and be able to quickly switch back and forth.
> 
> Would this change be possible?

I found a way to make this change. Added the PetsPage as a shell content. Navigating to the page via specific route keeps the tabs on screen and since the only way to access the pets page is via the social page; the back button just redirects to the social page. Also as bonus, navigating to timer and then back to social tab keeps the PetsPage open until the back button is hit.

iggy808 reviewed (approved):
LGTM

ChrispyPeaches reviewed (approved):
LGTM!

iggy808 reviewed (approved):
LETS GOOOOOOOOOOOOOOOOOOOOOOO
